(5) Cookie Banner

pezholio opened the pull request:
This adds a Cookie Banner to the site that enables all cookie-setting JavaScript (in our case Google Analytics) only if the user accepts.

It assumes that all cookie-using JS has a `data-cookies` element set and a `type="text/plain”` attribute (that prevents the JS from running).

The code then checks for the presence of a `hasConsent` cookie and  shows a banner if it has not been set. When a user hits `accept` the cookie is set.

If the cookie is set, all JS with a `data-cookies` attribute is activated by copying the script, and setting a type of `text/javascript`, which allows it to load.

This approach is lovingly nicked from https://github.com/nhsuk/cookie-consent

There's also a bit of extra Rollup/Babel config that I've added to make sure the JS works in all browsers (including IE11)

## Screenshots

### Cookie Banner

![image](https://user-images.githubusercontent.com/109774/165077401-dcf9cd2a-dce3-4246-8613-a2b1adfb95c1.png)

### Cookies Page

![image](https://user-images.githubusercontent.com/109774/165080382-d6cda5c2-d859-4817-91fc-fb807127d3a8.png)

pezholio committed:
Add a cookie banner script

This assumes that all cookie-using JS has a `data-cookies`
element set and a `type="text/plain”` attribute (that prevents
the JS from running).

The code then checks for the presence of a `hasConsent`
cookie and  shows a banner if it has not been set. When a user
hits `accept` the cookie is set.

If the cookie is set, all JS with a `data-cookies` attribute is
activated by copying the script, and setting a type of
`text/javascript`, which allows it to load.

This approach is lovingly nicked from https://github.com/nhsuk/cookie-consent

pezholio committed:
Fix eslint config

pezholio committed:
Add a Cookie banner

This is hidden by default, but we’ll add some JS to show this
in the next commit

pezholio committed:
Add a Cookies page

CristinaRO commented on app/views/layouts/_cookie_banner.html.erb:
```diff
@@ -0,0 +1,6 @@
+<div id="cookieBanner" class="hidden">
+    By continuing to visit this site, you accept the use of cookies by Google Analytics for statistical purposes.
```
Can we make it so we don't collect stats if the user continues to use the site without selecting "Accept"?

I'm not a fan of sites that take ignoring the button as implicit consent.

If this is already the case, can we rephrase the part that implies data is being collected just by "continuing to visit this site"?

pezholio replied:
Whoops. That was a bad copy/paste decision. I agree, that's a Very Bad approach, and not one that's reflected in the code.

pezholio committed:
Move the canonical redirect to top of the routes

Now we have a cookies page (as reflected in the tests), it’s
revealed that the redirect only takes place if the route does not
exist (which then resolves to a 404). For the canonical redirect
to take place, we need it to be on all routes.

pezholio committed:
Add Babel / Rollup config

This allows us to use modern JS, but maintain compatibility
with older browsers (mainly targetting IE11)

pezholio committed:
Add the Google Analytics script

This has a `data-cookies` element set and a script type of
`text/plain`, so only runs if cookies have been accepted

CristinaRO commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Explore cookies banner / plausible](https://trello.com/c/qt2UwfiL/5-explore-cookies-banner-plausible)

CristinaRO commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Cookie page](https://trello.com/c/rMmUXwgj/21-cookie-page)

CristinaRO commented on app/views/layouts/_cookie_banner.html.erb:
```diff
@@ -0,0 +1,7 @@
+<div id="cookieBanner" class="hidden">
+    <p>We use some essential cookies to make this service work.</p>
+    <p>We'd also like to use analytics cookies so we can understand how you use the service and make improvements.</p>    <a href="/cookies" id="cookies-eu-more">Read more</a>
+    <button id="rejectCookies">Reject</button>
```
What's the difference between "Reject" and "Reject analytics cookies" in practical terms? Are we able to not set `session[:uuid]` and `session[:just_rated_place_id]`, and do want to offer this as an option?

And what about the cookie that remembers they rejected cookies? (😬 )

pezholio replied:
Whoops - that first bit was a rebase gone bad. I'll need to add the `just_rated_place_id` to the strictly necessary cookies on the cookie page.

CristinaRO reviewed (approved):
Great work, thank you!
